pghoard/webserver.py

```diff
@@ -353,34 +390,46 @@ def get_wal_or_timeline_file(self, site, filename, filetype):
-            except (ValueError, HttpResponse) as ex:
+            ex = self._try_save_and_verify_restored_file(filetype, filename, xlog_path, target_path, unlink=False)
+            if ex:
                 self.server.log.warning("Found file: %r but it was invalid: %s", xlog_path, ex)
-            if success:
+            else:
                 raise HttpResponse(status=201)
 
-        prefetch_n = self.server.config["restore_prefetch"]
-        prefetch = []
-        if filetype == "timeline":
-            tli_num = int(filename.replace(".history", ""), 16)
-            for _ in range(prefetch_n):
-                tli_num += 1
-                prefetch.append("{:08X}.history".format(tli_num))
-        elif filetype == "xlog":
-            xlog_num = int(filename, 16)
-            for _ in range(prefetch_n):
-                if xlog_num & 0xFF == 0xFF:
-                    xlog_num += 0xFFFFFF00
-                xlog_num += 1
-                prefetch.append("{:024X}".format(xlog_num))
+        key = self._make_file_key(site, filetype, filename)
+        with suppress(ValueError):
+            self.server.prefetch_404.remove(key)
+        self._create_fetch_operation(key, site, filetype, filename, max_age=5, suppress_error=False)
+        self._create_prefetch_operations(site, filetype, filename)
 
-        with self._prefetch(site, filetype, prefetch):
-            self._transfer_agent_op(site, filename, filetype, "DOWNLOAD", target_path=target_path)
-
-        raise HttpResponse(status=201)
+        last_schedule_call = time.monotonic()
+        start_time = time.monotonic()
+        retries = 2
+        while time.monotonic() - start_time <= 30:
+            self._process_completed_download_operations()
+            with self.server.lock:
+                if os.path.isfile(prefetch_target_path):
+                    ex = self._try_save_and_verify_restored_file(filetype, filename, prefetch_target_path, target_path)
+                    if not ex:
+                        raise HttpResponse(status=201)
+                    elif ex and retries == 0:
+                        raise ex  # pylint: disable=raising-bad-type
+                    retries -= 1
+            if key in self.server.prefetch_404:
+                raise HttpResponse(status=404)
+            with self.server.lock:
+                if key not in self.server.pending_download_ops:
+                    if retries == 0:
+                        raise HttpResponse(status=500)
+                    retries -= 1
+                    self._create_fetch_operation(key, site, filetype, filename, suppress_error=False)
+            if time.monotonic() - last_schedule_call >= 1:
+                last_schedule_call = time.monotonic()
+                # Replace existing download operation if it has been executing for too long
+                self._create_fetch_operation(key, site, filetype, filename, max_age=10, suppress_error=False)
+            time.sleep(0.001)
```

Comment:
Looping pretty hard?

Reply:
Removed the sleep altogether, it was quite useless to begin with. Replaced with blocking `queue.get()` call with 10 times longer wait time.